### PR TITLE
Implement tagging system

### LINKED ISSUE #16: Have a samplesheet with an extra level of grouping
### Description of feature

I would love to be able to add an extra level for grouping:
- group/project/cohort
  - individual/patient
     - sample (tumor/normal)
 
That way I would be able to have multiple samples from the same patients and differentiate them from samples from another patient within the same group/project/cohort.

### PR DESCRIPTION
**Pull request by Aratz**
This PR implements the tagging system described in #16.

I have one question before this is merged though. Before this PR, the `id` of a sample was made of the sample name, the and the group. This PR removes the lane and group fields, so the `id` now only contains the sample name. I'm pretty sure we sometime sequence the same sample on different lane, or in different runs. Is there anything we can do to ensure the sample names are unique in this case? Or should it be the responsibility of the pipeline users to give unique names to the samples? @kedhammar do you have any opinion on this?

Closes #16 

## PR checklist

- [X] This comment contains a description of changes (with reason).
- [X] If necessary, also make a PR on the nf-core/seqinspector _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [X] Make sure your code lints (`nf-core lint`).
- [X] Ensure the test suite passes (`nf-test test main.nf.test -profile test,docker`).
- [X] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [X] Usage Documentation in `docs/usage.md` is updated.
- [X] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [X] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: assets/schema_input.json
Comment (kedhammar) [inline assets/schema_input.json:39]: @Aratz maybe something like this?

**Comment by kedhammar** [comment]
> I have one question before this is merged though. Before this PR, the id of a sample was made of the sample name, the `lane(?)` and the group. This PR removes the lane and group fields, so the id now only contains the sample name. I'm pretty sure we sometime sequence the same sample on different lane, or in different runs. Is there anything we can do to ensure the sample names are unique in this case? Or should it be the responsibility of the pipeline users to give unique names to the samples? @kedhammar do you have any opinion on this?

@Aratz I like the idea of adhering to a kind of `tidy data` philosophy ensuring each row of the samplesheet corresponds to the lowest-level component we want to wrangle. Currently this is a demuxed `.fq` file or file pair, which we in turn assume corresponds to a single sample.

This is not true for e.g. a case where a sample has been ran multiple times, or has been prepped with multiple different indices in the same run.

I guess I see two alternatives:
1. We leave it to the user to make sure sample names are unique, and can add an assertion to enforce this.
2. We allow sample name collision, and treat the sample name kind of like a tag. We can instead construct a unique identifier for each samplesheet row from scratch, as long as the rows themselves has no overlap in their `.fq` paths.

File: README.md
Comment (MatthiasZepper) [inline README.md:48]: I prefer avoiding `patient` in the examples, since it suggests an application in a clinical setting and this pipeline is evidently no approved in-vitro diagnostic product or medical device.

File: assets/schema_input.json
Comment (MatthiasZepper) [inline assets/schema_input.json:40]: Why did you opt for exactly this specification? 

We basically need the quotes, because we ask for the same delimiter for the tags and the sample sheet. This may still cause issues when parsing. (_e.g. Excel in some versions auto-quotes all text fields_) and will be quite confusing for the users as well?

Is there any reason not to use e.g. a semi-colon or a colon as separator? Something like `project123:batch_B:location3` might work as well, no?

File: docs/usage.md
Comment (MatthiasZepper) [inline docs/usage.md:34]: This example CSV lacks the separating commas, no?

File: subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf
Comment (MatthiasZepper) [inline subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf:87]:
```suggestion
                def tags = meta.tags ? meta.tags.tokenize(",").trim().toLowerCase() : []
```
Not sure if we should enforce lower case in tags, but we are using the tags as part of the file name and some file systems, e.g. that of MacOS, are case-insensitive. So either we enforce the conversion to make sure that we do not accidentally overwrite files or we need to have a check function that searches for identical tags like "`batch3,Batch3`" and raises an error.

**Review by MatthiasZepper — COMMENTED** [review]
Nice work! I would, however, opt for a different input specification, and we should somehow consider case-insensitivity as well.

File: assets/schema_input.json
Comment (kedhammar) [reply assets/schema_input.json:40]: If using e.g. a colon as a tag separator would help us get rid of the quotes, I'd be in favor!

File: subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf
Comment (kedhammar) [reply subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf:87]: > or we need to have a check function that searches for identical tags like batch3 & Batch3 and raises an error.

This seems like the nicest approach to me, since we are likely to want to retain tags like `P12345` as uppercase at NGI at least.

Preventing the possibility of overwriting things, while still maintaining the user's preference of tag names seems like a good compromise.

File: assets/schema_input.json
Comment (Aratz) [reply assets/schema_input.json:40]: I'll change to columns :+1:

File: docs/usage.md
Comment (Aratz) [reply docs/usage.md:34]: It is mentioned above that this is shown as `.tsv` for readability

File: subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf
Comment (MatthiasZepper) [inline subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf:91]: What is the rationale to attach the file path(s) to the `meta.id` from the sample sheet?

File: subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf
Comment (MatthiasZepper) [inline subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf:122]: That is nifty! Well done!

**Review by MatthiasZepper — COMMENTED** [review]
I like, where this is going!

File: subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf
Comment (Aratz) [reply subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf:91]: This is related to the question I had about making the ids unique. If a sample is run e.g. on two different lanes, two rows will have the same id. This is my implementation of @kedhammar 's second alternative (see above: https://github.com/nf-core/seqinspector/pull/20#issuecomment-2312738755)

File: subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf
Comment (Aratz) [reply subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf:122]: Thanks!

File: assets/schema_input.json
Comment (MatthiasZepper) [reply assets/schema_input.json]: Actually, the two regexes are equivalent. My bad. 

I put the mandatory tag to the left and allowed it to be extended by further tags, and the original regex enforced one valid tag that may be preceded by 0 or more other.

File: subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf
Comment (kedhammar) [reply subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf:122]: Note that from 8779909fe25a35950a999847d0b052b1f188de9c, we assert to only allow lowercase letters in tags when the samplesheet is parsed. We can make it more lenient if we want though.

File: subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf
Comment (kedhammar) [reply subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf:91]: @MatthiasZepper we are kind of heading to a direction where the fastq path is the only unique property of each row, so I thought it made sense to make it the dataframe index, if you will.

File: assets/schema_input.json
Comment (kedhammar) [reply assets/schema_input.json]: @Aratz we've realized the regexes are equivalent, but feel free to commit the change if you find it more intuitive

Comment (Aratz) [reply assets/schema_input.json]: Are you certain about that? empty tags and trailing colons wouldn't match with the second half of the regex. Or am I missing something?

I tried the following in RegExr: https://regexr.com/85g21

**Comment by kedhammar** [comment]
I can run this (`nextflow run . -profile test,docker --outdir outdir`) on GitPod, but not locally (Mac M1), is this a problem that needs addressing?

**Comment by Aratz** [comment]
> I can run this (`nextflow run . -profile test,docker --outdir outdir`) on GitPod, but not locally (Mac M1), is this a problem that needs addressing?

What error message do you get? Is this something you get for other pipelines too? I think it would be better if this could be fixed as it will degrade the developer's experience for mac users :/

File: assets/schema_input.json
Comment (kedhammar) [reply assets/schema_input.json:39]: - The first regex is an optional series of colon-trailing tags followed by a tag
- The second regex is a tag followed by an optional series of colon-headed tags

At least I found them to be equivalent and that neither accept `:tag` or `tag:`

**Comment by kedhammar** [comment]
> What error message do you get?
```
Sep-03 17:06:47.486 [Task monitor] DEBUG nextflow.processor.TaskProcessor - Handling unexpected condition for
  task: name=NFCORE_SEQINSPECTOR:SEQINSPECTOR:MULTIQC_GLOBAL; work-dir=/Users/alfred.kedhammar/code/repos/seqinspector/work/10/a072373ac544f0e1ba539172e09295
  error [nextflow.exception.ProcessFailedException]: Process `NFCORE_SEQINSPECTOR:SEQINSPECTOR:MULTIQC_GLOBAL` terminated with an error exit status (1)
Sep-03 17:06:47.532 [Task monitor] ERROR nextflow.processor.TaskProcessor - Error executing process > 'NFCORE_SEQINSPECTOR:SEQINSPECTOR:MULTIQC_GLOBAL'

Caused by:
  Process `NFCORE_SEQINSPECTOR:SEQINSPECTOR:MULTIQC_GLOBAL` terminated with an error exit status (1)

Command executed:

  multiqc \
      --force \
       \
      --config multiqc_config.yml \
       \
       \
       \
       \
       \
      .
  
  cat <<-END_VERSIONS > versions.yml
  "NFCORE_SEQINSPECTOR:SEQINSPECTOR:MULTIQC_GLOBAL":
      multiqc: $( multiqc --version | sed -e "s/multiqc, version //g" )
  END_VERSIONS

Command exit status:
  1

Command output:
  (empty)

Command error:
  #17 0x55555a36c874 base::Thread::ThreadMain()
  #18 0x55555a38ecaa base::(anonymous namespace)::ThreadFunc()
  #19 0x2aaaab656134 (/usr/lib/x86_64-linux-gnu/libc.so.6+0x89133)
  #20 0x2aaaab6d5a40 __clone
  Task trace:
  #0 0x555558d5905e content::internal::ChildProcessLauncherHelper::PostLaunchOnLauncherThread()
  #1 0x555558d58afc content::internal::ChildProcessLauncherHelper::StartLaunchOnClientThread()
  #2 0x55555928a7c2 content::VizProcessTransportFactory::ConnectHostFrameSinkManager()
  #3 0x55555a9ae26f mojo::SimpleWatcher::Context::Notify()
  #4 0x55555928a7c2 content::VizProcessTransportFactory::ConnectHostFrameSinkManager()
  Task trace buffer limit hit, update PendingTask::kTaskBacktraceLength to increase.
  
  Received signal 6
  #0 0x55555a37ed79 base::debug::CollectStackTrace()
  #1 0x55555a2fc633 base::debug::StackTrace::StackTrace()
  #2 0x55555a37e95b base::debug::(anonymous namespace)::StackDumpSignalHandler()
  #3 0x2aaaab609050 (/usr/lib/x86_64-linux-gnu/libc.so.6+0x3c04f)
  #4 0x2aaaab657e2c (/usr/lib/x86_64-linux-gnu/libc.so.6+0x8ae2b)
  #5 0x2aaaab608fb2 gsignal
  #6 0x2aaaab5f3472 abort
  #7 0x55555a37d8f5 base::debug::BreakDebugger()
  #8 0x55555a30c5b2 logging::LogMessage::~LogMessage()
  #9 0x555558e68907 content::(anonymous namespace)::IntentionallyCrashBrowserForUnusableGpuProcess()
  #10 0x555558e666be content::GpuDataManagerImplPrivate::FallBackToNextGpuMode()
  #11 0x555558e64c53 content::GpuDataManagerImpl::FallBackToNextGpuMode()
  #12 0x555558e6ef5f content::GpuProcessHost::RecordProcessCrash()
  #13 0x555558cf96d3 content::BrowserChildProcessHostImpl::OnProcessLaunchFailed()
  #14 0x555558d591f0 content::internal::ChildProcessLauncherHelper::PostLaunchOnClientThread()
  #15 0x555558d59435 base::internal::Invoker<>::RunOnce()
  #16 0x55555a342306 base::TaskAnnotator::RunTask()
  #17 0x55555a353cf6 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWorkImpl()
  #18 0x55555a3539ea base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWork()
  #19 0x55555a3a0899 base::MessagePumpLibevent::Run()
  #20 0x55555a35459b base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::Run()
  #21 0x55555a32d9bd base::RunLoop::Run()
  #22 0x555558d15f18 content::BrowserProcessSubThread::IOThreadRun()
  #23 0x55555a36c874 base::Thread::ThreadMain()
  #24 0x55555a38ecaa base::(anonymous namespace)::ThreadFunc()
  #25 0x2aaaab656134 (/usr/lib/x86_64-linux-gnu/libc.so.6+0x89133)
  #26 0x2aaaab6d5a40 __clone
    r8: 0000000000000000  r9: 0000000000000000 r10: 0000000000000008 r11: 00000000ae316e43
   r12: 0000000000000006 r13: 00002aaaae318580 r14: 00002aaaae318590 r15: 00002aaaae318598
    di: 0000000000000610  si: 000000000000061d  bp: 00002aaaae3196c0  bx: 000000000000061d
    dx: 0000000000000006  ax: 0000000000000000  cx: 0000000000000000  sp: 00002aaaae3173c0
    ip: 00002aaaab657e2c efl: 0000000000000246 cgf: 002b000000000033 erf: 0000000000000000
   trp: ffffffffffffffff msk: 0000000000000000 cr2: 0000000000000000
  [end of stack trace]
```

>Is this something you get for other pipelines too? 

It works when I run on `upstream/dev` so it's something within this branch I guess.

**Comment by Aratz** [comment]
Can you try from `upstream/dev` after updating multiqc?

File: subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf
Comment (MatthiasZepper) [reply subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf:91]: Ah, ok, that is indeed a valid concern. But can't we just enumerate the lines of the sample sheet or something? `fastq_1.toString().replaceAll('/', '_')` feels quite massive...

Comment (MatthiasZepper) [reply subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf:91]: Maybe not deterministic...because the order of the channel may change with a real `.fromSamplesheet()` factory?!? 

```
    ch_fruit = Channel.from("apple","mango", "banana", "banana","peach")
    ch_fruit_unique = ch_fruit.toList().flatMap { it.withIndex().collect{ entry, idx -> "${entry}_${idx}" }}
    ch_fruit_unique.view()
```
which prints
```
apple_0
mango_1
banana_2
banana_3
peach_4
```

Comment (MatthiasZepper) [reply subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf:91]: I would prefer a simple number suffix being added to the `meta.sample` to form the `meta.id` :

```

Channel.fromSamplesheet("input")
    .toList()
    .flatMap{it.withIndex().collect{ entry, idx -> entry + "${idx+1}" }}
    .map { meta, fastq_1, fastq_2, idx -> 
        def tags = meta.tags ? meta.tags.tokenize(":") : []
        def updated_meta = meta + [ id: "${meta.sample}_${idx}", tags:tags, single: fastq_2[0] ? true : false ] 
        return [updated_meta.id, updated_meta, fastq_1 , fastq_2]
    }.view()
```

**Comment by kedhammar** [comment]
> Can you try from `upstream/dev` after updating multiqc?

I tried it now and appear to get the same error. So the change that made it stop crashing on GitPod made it start crashing locally I guess 😆

**Comment by kedhammar** [comment]
I was able to replicate the crash on @MatthiasZepper 's computer and have found some old issues on the MultiQC repo indicating the problem is not on our end.

I've made an issue, as shown above https://github.com/MultiQC/MultiQC/issues/2812

**Comment by kedhammar** [comment]
My issue has been replied to and this is indeed more a problem on MultiQCs end than our, 
see https://github.com/MultiQC/MultiQC/issues/2812#issuecomment-2332011779

I'm fine with going ahead now, but if we would rather crash GitPod than ARM Macs, we could rollback to MultiQC v1.21. Personally I prefer prioritizing functionality on GitPod.

**Review by kedhammar — APPROVED** [review]

**Comment by FranBonath** [comment]
Hej @Aratz @kedhammar, is there still more work to be done on this PR, or can it be merged?

**Comment by MatthiasZepper** [comment]
> Hej @Aratz @kedhammar, is there still more work to be done on this PR, or can it be merged?

I am ok with merging, but don't find using the file path as ID particularly elegant or helpful (also because it will be displayed when running the pipeline by Nextflow ?!?). Thus I proposed that, but haven't gotten any feedback here. 

```
Channel.fromSamplesheet("input")
    .toList()
    .flatMap{it.withIndex().collect{ entry, idx -> entry + "${idx+1}" }}
    .map { meta, fastq_1, fastq_2, idx -> 
        def tags = meta.tags ? meta.tags.tokenize(":") : []
        def updated_meta = meta + [ id: "${meta.sample}_${idx}", tags:tags, single: fastq_2[0] ? true : false ] 
        return [updated_meta.id, updated_meta, fastq_1 , fastq_2]
    }
    .set { ch_samplesheet }
```

File: subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf
Comment (Aratz) [reply subworkflows/local/utils_nfcore_seqinspector_pipeline/main.nf:91]: I've asked for guidance on the nf-core slack (https://nfcore.slack.com/archives/C043FMKUNLB/p1725872244960039), from what I understand `id` can be useful to track down samples when debugging. So I'd follow James' suggestion and just use filenames instead of full paths.

**Review by MatthiasZepper — APPROVED** [review]
Sorry, I was on vacation, but since you already had the approval from Alfred, I assumed you would merge right away! I am fine with merging as well, in particular since I persuaded you about the IDs 😉

**Comment by Aratz** [comment]
No worries, thanks for the great feedback!

**Pull request merged**
